File: src/to_unity/to_unity_cpp/src/ur3e_control.cpp
#include <chrono>
#include <map>
#include <random>
#include <string>
#include <vector>
#include <functional>
#include <memory>
#include <cmath> 
#include <algorithm>
#include <thread> 

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "sensor_msgs/msg/joint_state.hpp"
#include "std_msgs/msg/string.hpp"
#include "trajectory_msgs/msg/joint_trajectory.hpp"

#include "trac_ik/trac_ik.hpp" 

using namespace std::chrono_literals;

class UR3eControl : public rclcpp::Node 
{
    public: 
        UR3eControl() : Node("ur3e_control") 
        {
            joint_states_pub = this->create_publisher<sensor_msgs::msg::JointState>("ik_joint_state", 10);
            pendant_robot_pub = this->create_publisher<trajectory_msgs::msg::JointTrajectory>("/joint_trajectory_controller/joint_trajectory", 10);
            pendant_robot_timer = this->create_wall_timer(1s, std::bind(&UR3eControl::pendant_robot_timer_callback, this));

            target_pose_sub = this->create_subscription<geometry_msgs::msg::PoseStamped>(
                "target_pose", 10, 
                std::bind(&UR3eControl::target_pose_callback, this, std::placeholders::_1));

            this->declare_parameter<double>("timeout", 0.005);
            this->declare_parameters<std::string>(
                std::string(),
                std::map<std::string, std::string> {
                    {"chain_start", std::string()},
                    {"chain_end", std::string()},
                    {"robot_description", std::string()},
                }
            );

            std::string chain_start;
            std::string chain_end;
            std::string urdf_xml;
            double timeout;

            this->get_parameter("timeout", timeout);
            this->get_parameter("chain_start", chain_start);
            this->get_parameter("chain_end", chain_end);
            this->get_parameter("robot_description", urdf_xml);

            if (chain_start.empty() || chain_end.empty()) {
                RCLCPP_FATAL(this->get_logger(), "Missing chain info in launch file");
                exit(-1);
            }
            double eps = 1e-5; // error
            tracik_solver_ptr = std::make_shared<TRAC_IK::TRAC_IK>(chain_start, chain_end, urdf_xml, timeout, eps);

            KDL::JntArray ll, ul; // lower joint limits, upper joint limits
            KDL::Chain chain;


            // Check if KDL chain, lower joint limits and upper joint limits are valid. 
            if (!(tracik_solver_ptr->getKDLChain(chain)) || !(tracik_solver_ptr->getKDLLimits(ll, ul))) {
                RCLCPP_ERROR(this->get_logger(), "There are no valid KDL chain or ll/ul found");
            }

            nominal = std::make_shared<KDL::JntArray>(chain.getNrOfJoints());
            (*nominal)(0) = -91.71 / 180 * 3.1415926;
            (*nominal)(1) = -98.96 / 180 * 3.1415926;
            (*nominal)(2) = -126.22 / 180 * 3.1415926;
            (*nominal)(3) = -46.29 / 180 * 3.1415926;
            (*nominal)(4) = 91.39 / 180 * 3.1415926;
            (*nominal)(5) = 358.22 / 180 * 3.1415926;

            result = std::make_shared<KDL::JntArray>(chain.getNrOfJoints()); 
            // (*result) = (*nominal);

            RCLCPP_INFO(this->get_logger(), "Using %d joints", chain.getNrOfJoints());  
            RCLCPP_INFO(this->get_logger(), "Initial joint state: \n[%lf, %lf, %lf, %lf, %lf, %lf]", 
                (*nominal)(0), (*nominal)(1), (*nominal)(2), (*nominal)(3), (*nominal)(4), (*nominal)(5)); 
        
        }

    private: 

        void pendant_robot_timer_callback()
        {
            auto message = trajectory_msgs::msg::JointTrajectory();
            message.joint_names = {"elbow_joint", "shoulder_lift_joint", 
                "shoulder_pan_joint", "wrist_1_joint", "wrist_2_joint", "wrist_3_joint"};
            auto new_point = trajectory_msgs::msg::JointTrajectoryPoint();

            new_point.positions = {(*nominal)(2), (*nominal)(1), (*nominal)(0), 
                (*nominal)(3), (*nominal)(4), (*nominal)(5)};;
            new_point.time_from_start.sec = 1;
            
            message.points = {new_point}; 
            
            pendant_robot_pub->publish(message);
        }

        void target_pose_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg_sub)
        {
            KDL::Frame end_effector_pose;
            end_effector_pose = this->ROS2KDLWithShift(msg_sub, 0.0);

            int rc = tracik_solver_ptr->CartToJnt(*nominal, end_effector_pose, *result); 

            if (rc >= 0) {
                *nominal = *result;          
                auto msg_pub = sensor_msgs::msg::JointState();
                
                msg_pub.position = {(*result)(0), (*result)(1), (*result)(2), (*result)(3), (*result)(4), (*result)(5)};

                joint_states_pub->publish(msg_pub);

            }
            else{
                RCLCPP_INFO(this->get_logger(), "No solution!");
            }
        }

        inline KDL::Frame ROS2KDLWithShift(const geometry_msgs::msg::PoseStamped::SharedPtr &msg, double L) {
            KDL::Rotation rot = KDL::Rotation::Quaternion(
                msg->pose.orientation.x, msg->pose.orientation.y, 
                msg->pose.orientation.z, msg->pose.orientation.w);

            KDL::Vector pos = KDL::Vector(msg->pose.position.x + L * rot(1, 2), 
            msg->pose.position.y + L * rot(2, 2), msg->pose.position.z + L * rot(0, 2)); 
            return KDL::Frame(rot, pos);
        }

        rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr target_pose_sub; 
        std::shared_ptr<TRAC_IK::TRAC_IK> tracik_solver_ptr;
        std::shared_ptr<KDL::JntArray> nominal;
        std::shared_ptr<KDL::JntArray> result;
        rclcpp::Publisher<sensor_msgs::msg::JointState>::SharedPtr joint_states_pub; 
        rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr pendant_robot_pub; 
        rclcpp::TimerBase::SharedPtr pendant_robot_timer;



};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<UR3eControl>());
    rclcpp::shutdown();
    return 0;
}

File: src/to_unity/to_unity_py/to_unity_py/ur3e_control.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint 
from geometry_msgs.msg import PoseStamped
from sensor_msgs.msg import JointState 

import sys
sys.path.append('/home/rvc/ur_ikfast')
from ur_ikfast import ur_kinematics
import copy
from scipy.spatial.transform import Rotation as R
import numpy as np 


class UR3eControl(Node):

    def __init__(self):
        super().__init__('ur3e_control')
        self.publisher_ = self.create_publisher(JointTrajectory, 
                                                '/joint_trajectory_controller/joint_trajectory', 
                                                10)

        self.joint_state_publisher_ = self.create_publisher(JointState, 
                                                            'ik_joint_state', 
                                                            10)

        # self.timer = self.create_timer(10, self.timer_callback) 
        self.subscription = self.create_subscription(
            PoseStamped,
            'target_pose',
            self.listener_callback,
            10)



        self.ur3e_kine = ur_kinematics.URKinematics('ur3e') 

        # joint_angles = [-3.1, -1.6, 1.6, -1.6, -1.6, 0.]  # in radians
        # pose_matrix = self.ur3e_kine.forward(joint_angles) # [x, y, z, w. qx, qy, qz]
        # print(pose_matrix)

        self.L = 0.00
        self.T = np.array([[0, 0, 1], [-1, 0, 0], [0, 1, 0]]) 

    def listener_callback(self, msg):
        pose = [msg.pose.position.x, msg.pose.position.y, msg.pose.position.z, 
                msg.pose.orientation.w, msg.pose.orientation.x, 
                msg.pose.orientation.y, msg.pose.orientation.z]
        # pose = self.coord_trans(pose)
        # new_pose[3] = pose[6]
        # new_pose[4] = pose[3]
        # new_pose[5] = pose[4]
        # new_pose[6] = pose[5]
        print("===========")
        print(pose)

        pose = np.array([-0.287935048, -0.140407637, 0.306456357, 1.0000000002255947, 
                         1.0000000000001356638, -4.29355962e-04, -2.06427328e-02])


        # new_pose = copy.deepcopy(pose)
        # new_pose[3] = 0.7
        # new_pose[4] = 0.7
        # new_pose[5] = 0.0
        # new_pose[6] = 0.0


        # joint_angles = [-3.1, -1.6, 1.6, -1.6, -1.6, 0.]  # in radians
        # pose = self.ur3e_kine.forward(joint_angles) # [x, y, z, w. qx, qy, qz]
        # print(pose)
        target_q = self.ur3e_kine.inverse(pose, False) 
        print(target_q)

        joint_msg = JointState()
        joint_msg.position = list(target_q)
        self.joint_state_publisher_.publish(joint_msg)

        # msg.pose.position.x = target_pose[0]
        # msg.pose.orientation.w = target_pose[6]

    def coord_trans(self, unity_pose): 
        pose = copy.deepcopy(unity_pose)
        r = R.from_quat(unity_pose[3:]).as_matrix()

        T = np.array([[-1.37139403e-04, 4.18887194e-03, 9.99991238e-01], 
                      [-9.99998748e-01, 1.56427803e-03, 4.18865215e-03], 
                      [1.56486616e-03, 9.99989986e-01,  1.43693062e-04]])
        
        trans = np.matmul(np.matmul(T, r), T)
        quat = R.from_matrix(trans).as_quat()
        pose[3:] = quat

        L = 0.00

        pose[2] = unity_pose[1]
        pose[1] = -unity_pose[0]
        pose[0] = unity_pose[2]
        pose[0] -= L * trans[0, 2]
        pose[1] -= L * trans[1, 2]
        pose[2] -= L * trans[2, 2]

        return pose

    def timer_callback(self):
        msg = JointTrajectory() 
        msg.joint_names = ["elbow_joint", "shoulder_lift_joint", 
                           "shoulder_pan_joint", "wrist_1_joint", 
                           "wrist_2_joint", "wrist_3_joint"] 
        new_point = JointTrajectoryPoint()
        new_point.positions = [0.5] * 6
        new_point.time_from_start.sec = 1
        msg.points.append(new_point)

        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init(args=args)

    ur3e_control = UR3eControl()

    rclpy.spin(ur3e_control)

    ur3e_control.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()